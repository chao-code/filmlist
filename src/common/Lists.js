import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Link } from 'react-router-dom'
import Poster from './Poster'

const renderList = (list, compact) => {
  const { listID, posters, subject, listedBy, filmCount, description } = list
  const url = `/list/${listID}`
  const row = classNames({ row: !compact })
  const col4 = classNames({ 'col-md-4': !compact })
  const col8 = classNames({ 'col-md-8': !compact })
  const text = `${filmCount} ${filmCount === 1 ? 'film' : 'films'} listed by ${listedBy}`
  const offset = description.indexOf('\n')
  const brief = offset < 0 ? description : description.slice(0, offset)

  return (
    <li className="my-3" key={listID}>
      <div className={row}>
        <div className={col4}>
          <div className="row no-gutters mb-2">
            {[0, 1, 2, 3, 4].map(i => (
              <div className="col" key={i}>
                {posters[i] &&
                  <Link to={url}>
                    <Poster poster={posters[i]} crop />
                  </Link>
                }
              </div>
            ))}
          </div>
        </div>
        <div className={col8}>
          {compact ? (
            <h6 className="mb-0">
              <Link to={url} className="link">{subject}</Link>
            </h6>
          ) : (
            <h5 className="mb-0">
              <Link to={url} className="link">{subject}</Link>
            </h5>
          )}
          <small className="text-muted">{text}</small>
          {!compact &&
            <div>{brief}</div>
          }
        </div>
      </div>
    </li>
  )
}

const Lists = ({ lists, compact }) => (
  <ul className="list-unstyled">
    {lists.map(list => renderList(list, compact))}
  </ul>
)

Lists.propTypes = {
  lists: PropTypes.arrayOf(PropTypes.shape({
    listID: PropTypes.string,
    posters: PropTypes.array,
    listedBy: PropTypes.string,
    filmCount: PropTypes.number,
    description: PropTypes.string,
  })).isRequired,
  compact: PropTypes.bool,
}

Lists.defaultProps = {
  compact: false,
}

export default Lists
