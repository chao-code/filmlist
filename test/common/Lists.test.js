import { shallow } from 'enzyme'
import Lists from '../../src/common/Lists'
import { mockRecentLists as mockLists } from '../mocks/dataMock'

describe('<Lists />', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Lists lists={mockLists} />)
    expect(wrapper.find('li').length).toBe(mockLists.length)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render compact correctly', () => {
    const wrapper = shallow(<Lists lists={mockLists} compact />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should show the posters of each list', () => {
    const wrapper = shallow(<Lists lists={mockLists} />)
    const list = wrapper.find('li').first()
    expect(list.find('Poster').length).toBe(mockLists.length)
  })

  it('should link to a list path', () => {
    const wrapper = shallow(<Lists lists={mockLists} />)
    const list = wrapper.find('li').first()
    const path = list.find('Link').first().prop('to')
    expect(path).toBe(`/list/${mockLists[0].listID}`)
  })
})
