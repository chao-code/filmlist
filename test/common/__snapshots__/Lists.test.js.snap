// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lists /> should render compact correctly 1`] = `
<ul
  className="list-unstyled"
>
  <li
    className="my-3"
    key="22222"
  >
    <div
      className=""
    >
      <div
        className=""
      >
        <div
          className="row no-gutters mb-2"
        >
          <div
            className="col"
            key="0"
          >
            <Link
              replace={false}
              to="/list/22222"
            >
              <Poster
                crop={true}
                poster="1.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="1"
          >
            <Link
              replace={false}
              to="/list/22222"
            >
              <Poster
                crop={true}
                poster="2.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="2"
          >
            <Link
              replace={false}
              to="/list/22222"
            >
              <Poster
                crop={true}
                poster="3.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="3"
          />
          <div
            className="col"
            key="4"
          />
        </div>
      </div>
      <div
        className=""
      >
        <h6
          className="mb-0"
        >
          <Link
            className="link"
            replace={false}
            to="/list/22222"
          >
            List B
          </Link>
        </h6>
        <small
          className="text-muted"
        >
          3 films listed by user2
        </small>
      </div>
    </div>
  </li>
  <li
    className="my-3"
    key="33333"
  >
    <div
      className=""
    >
      <div
        className=""
      >
        <div
          className="row no-gutters mb-2"
        >
          <div
            className="col"
            key="0"
          >
            <Link
              replace={false}
              to="/list/33333"
            >
              <Poster
                crop={true}
                poster="3.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="1"
          >
            <Link
              replace={false}
              to="/list/33333"
            >
              <Poster
                crop={true}
                poster="1.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="2"
          />
          <div
            className="col"
            key="3"
          />
          <div
            className="col"
            key="4"
          />
        </div>
      </div>
      <div
        className=""
      >
        <h6
          className="mb-0"
        >
          <Link
            className="link"
            replace={false}
            to="/list/33333"
          >
            List C
          </Link>
        </h6>
        <small
          className="text-muted"
        >
          2 films listed by user2
        </small>
      </div>
    </div>
  </li>
  <li
    className="my-3"
    key="11111"
  >
    <div
      className=""
    >
      <div
        className=""
      >
        <div
          className="row no-gutters mb-2"
        >
          <div
            className="col"
            key="0"
          >
            <Link
              replace={false}
              to="/list/11111"
            >
              <Poster
                crop={true}
                poster="1.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="1"
          />
          <div
            className="col"
            key="2"
          />
          <div
            className="col"
            key="3"
          />
          <div
            className="col"
            key="4"
          />
        </div>
      </div>
      <div
        className=""
      >
        <h6
          className="mb-0"
        >
          <Link
            className="link"
            replace={false}
            to="/list/11111"
          >
            List A
          </Link>
        </h6>
        <small
          className="text-muted"
        >
          1 film listed by user1
        </small>
      </div>
    </div>
  </li>
</ul>
`;

exports[`<Lists /> should render correctly 1`] = `
<ul
  className="list-unstyled"
>
  <li
    className="my-3"
    key="22222"
  >
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
        <div
          className="row no-gutters mb-2"
        >
          <div
            className="col"
            key="0"
          >
            <Link
              replace={false}
              to="/list/22222"
            >
              <Poster
                crop={true}
                poster="1.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="1"
          >
            <Link
              replace={false}
              to="/list/22222"
            >
              <Poster
                crop={true}
                poster="2.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="2"
          >
            <Link
              replace={false}
              to="/list/22222"
            >
              <Poster
                crop={true}
                poster="3.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="3"
          />
          <div
            className="col"
            key="4"
          />
        </div>
      </div>
      <div
        className="col-md-8"
      >
        <h5
          className="mb-0"
        >
          <Link
            className="link"
            replace={false}
            to="/list/22222"
          >
            List B
          </Link>
        </h5>
        <small
          className="text-muted"
        >
          3 films listed by user2
        </small>
        <div />
      </div>
    </div>
  </li>
  <li
    className="my-3"
    key="33333"
  >
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
        <div
          className="row no-gutters mb-2"
        >
          <div
            className="col"
            key="0"
          >
            <Link
              replace={false}
              to="/list/33333"
            >
              <Poster
                crop={true}
                poster="3.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="1"
          >
            <Link
              replace={false}
              to="/list/33333"
            >
              <Poster
                crop={true}
                poster="1.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="2"
          />
          <div
            className="col"
            key="3"
          />
          <div
            className="col"
            key="4"
          />
        </div>
      </div>
      <div
        className="col-md-8"
      >
        <h5
          className="mb-0"
        >
          <Link
            className="link"
            replace={false}
            to="/list/33333"
          >
            List C
          </Link>
        </h5>
        <small
          className="text-muted"
        >
          2 films listed by user2
        </small>
        <div />
      </div>
    </div>
  </li>
  <li
    className="my-3"
    key="11111"
  >
    <div
      className="row"
    >
      <div
        className="col-md-4"
      >
        <div
          className="row no-gutters mb-2"
        >
          <div
            className="col"
            key="0"
          >
            <Link
              replace={false}
              to="/list/11111"
            >
              <Poster
                crop={true}
                poster="1.jpg"
              />
            </Link>
          </div>
          <div
            className="col"
            key="1"
          />
          <div
            className="col"
            key="2"
          />
          <div
            className="col"
            key="3"
          />
          <div
            className="col"
            key="4"
          />
        </div>
      </div>
      <div
        className="col-md-8"
      >
        <h5
          className="mb-0"
        >
          <Link
            className="link"
            replace={false}
            to="/list/11111"
          >
            List A
          </Link>
        </h5>
        <small
          className="text-muted"
        >
          1 film listed by user1
        </small>
        <div />
      </div>
    </div>
  </li>
</ul>
`;
